File: src/main/java/io/github/takenoha/towerdefense/persistence/TeamResourceSnapshot.java
package io.github.takenoha.towerdefense.persistence;

import java.util.Objects;
import java.util.UUID;

/** A database-derived team wallet view plus the viewer's active-event provisional claims. */
public record TeamResourceSnapshot(
        UUID teamId,
        long defensePoints,
        long enhancementPoints,
        long teamProvisionalDefensePoints,
        long teamProvisionalEnhancementPoints,
        long viewerProvisionalDefensePoints,
        long viewerProvisionalEnhancementPoints) {
    /** Compatibility constructor for callers that only supplied viewer provisional totals. */
    public TeamResourceSnapshot(
            UUID teamId,
            long defensePoints,
            long enhancementPoints,
            long provisionalDefensePoints,
            long provisionalEnhancementPoints) {
        this(
                teamId,
                defensePoints,
                enhancementPoints,
                provisionalDefensePoints,
                provisionalEnhancementPoints,
                provisionalDefensePoints,
                provisionalEnhancementPoints);
    }

    public TeamResourceSnapshot {
        Objects.requireNonNull(teamId, "teamId");
        if (defensePoints < 0 || enhancementPoints < 0
                || teamProvisionalDefensePoints < 0 || teamProvisionalEnhancementPoints < 0
                || viewerProvisionalDefensePoints < 0 || viewerProvisionalEnhancementPoints < 0) {
            throw new IllegalArgumentException("resource quantities must not be negative");
        }
    }

    public long balance(ResourceType type) {
        return switch (ResourceType.require(type)) {
            case DEFENSE_POINTS -> defensePoints;
            case ENHANCEMENT_POINTS -> enhancementPoints;
        };
    }

    public long provisional(ResourceType type) {
        return switch (ResourceType.require(type)) {
            case DEFENSE_POINTS -> viewerProvisionalDefensePoints;
            case ENHANCEMENT_POINTS -> viewerProvisionalEnhancementPoints;
        };
    }

    public long teamProvisional(ResourceType type) {
        return switch (ResourceType.require(type)) {
            case DEFENSE_POINTS -> teamProvisionalDefensePoints;
            case ENHANCEMENT_POINTS -> teamProvisionalEnhancementPoints;
        };
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/persistence/TeamResourceSnapshot.kt
package io.github.takenoha.towerdefense.persistence

import java.util.Objects
import java.util.UUID
import kotlin.jvm.JvmRecord

/** A database-derived team wallet view plus the viewer's active-event provisional claims. */
@JvmRecord
data class TeamResourceSnapshot(
    val teamId: UUID,
    val defensePoints: Long,
    val enhancementPoints: Long,
    val teamProvisionalDefensePoints: Long,
    val teamProvisionalEnhancementPoints: Long,
    val viewerProvisionalDefensePoints: Long,
    val viewerProvisionalEnhancementPoints: Long,
) {
    /** Compatibility constructor for callers that only supplied viewer provisional totals. */
    constructor(
        teamId: UUID,
        defensePoints: Long,
        enhancementPoints: Long,
        provisionalDefensePoints: Long,
        provisionalEnhancementPoints: Long,
    ) : this(
        teamId,
        defensePoints,
        enhancementPoints,
        provisionalDefensePoints,
        provisionalEnhancementPoints,
        provisionalDefensePoints,
        provisionalEnhancementPoints,
    )

    init {
        Objects.requireNonNull(teamId, "teamId")
        if (defensePoints < 0L || enhancementPoints < 0L ||
            teamProvisionalDefensePoints < 0L || teamProvisionalEnhancementPoints < 0L ||
            viewerProvisionalDefensePoints < 0L || viewerProvisionalEnhancementPoints < 0L
        ) {
            throw IllegalArgumentException("resource quantities must not be negative")
        }
    }

    fun balance(type: ResourceType): Long = when (ResourceType.require(type)) {
        ResourceType.DEFENSE_POINTS -> defensePoints
        ResourceType.ENHANCEMENT_POINTS -> enhancementPoints
    }

    fun provisional(type: ResourceType): Long = when (ResourceType.require(type)) {
        ResourceType.DEFENSE_POINTS -> viewerProvisionalDefensePoints
        ResourceType.ENHANCEMENT_POINTS -> viewerProvisionalEnhancementPoints
    }

    fun teamProvisional(type: ResourceType): Long = when (ResourceType.require(type)) {
        ResourceType.DEFENSE_POINTS -> teamProvisionalDefensePoints
        ResourceType.ENHANCEMENT_POINTS -> teamProvisionalEnhancementPoints
    }
}
